test/Microsoft.ML.Tests/Scenarios/Api/MultithreadedPrediction.cs

```diff
@@ -0,0 +1,157 @@
+        /// reallocate internal memory buffers on every single prediction, the PredictionEngine
+        /// (or its estimator/transformer based successor) is, like most stateful .NET objects,
+        /// fundamentally not thread safe. This is deliberate and as designed. However, some mechanism
+        /// to enable multi-threaded scenarios (e.g., a web server servicing requests) should be possible
+        /// and performant in the new API.
+        /// </summary>
+        [Fact]
+        void MultithreadedPrediction()
+        {
+            var dataPath = GetDataPath(SentimentDataPath);
+            var testDataPath = GetDataPath(SentimentTestPath);
+
+            using (var env = new TlcEnvironment(seed: 1, conc: 1))
+            {
+                // Pipeline
+                var loader = new TextLoader(env, MakeSentimentTextLoaderArgs(), new MultiFileSource(dataPath));
+
+                var trans = TextTransform.Create(env, MakeSentimentTextTransformArgs(), loader);
+
+                // Train
+                var trainer = new LinearClassificationTrainer(env, new LinearClassificationTrainer.Arguments
+                {
+                    NumThreads = 1
+                });
+
+                var cached = new CacheDataView(env, trans, prefetch: null);
+                var trainRoles = new RoleMappedData(cached, label: "Label", feature: "Features");
+                var predictor = trainer.Train(new Runtime.TrainContext(trainRoles));
+
+                var scoreRoles = new RoleMappedData(trans, label: "Label", feature: "Features");
+                IDataScorerTransform scorer = ScoreUtils.GetScorer(predictor, scoreRoles, env, trainRoles.Schema);
+
+                // Create prediction engine and test predictions.
+                var model = env.CreatePredictionEngine<SentimentData, SentimentPrediction>(scorer);
+
+                // Take a couple examples out of the test data and run predictions on top.
+                var testLoader = new TextLoader(env, MakeSentimentTextLoaderArgs(), new MultiFileSource(GetDataPath(SentimentTestPath)));
+                var testData = testLoader.AsEnumerable<SentimentData>(env, false);
+                var lockEngine = new LockBasedPredictionEngine(env, model);
+                Parallel.ForEach(testData, (input) => lockEngine.Predict(input));
+                int numThreads = 2;
+                PredictionEngine<SentimentData, SentimentPrediction>[] models = new PredictionEngine<SentimentData, SentimentPrediction>[numThreads];
+                using (var file = env.CreateTempFile())
+                {
+                    // Save model. 
+                    using (var ch = env.Start("saving"))
+                        TrainUtils.SaveModel(env, ch, file, predictor, scoreRoles);
+                    var threadEngine = new ThreadLocalBasedPredictionEngine(env, file);
+                    var poolEngine = new PoolBasedPredictionEngine(env, file);
```

Comment:
>poolEngine [](start = 24, length = 10)

let's get rid of this guy, and only keep the thread-local

Reply:
Why, they provide examples of how it can be handled in a different ways.

---
In reply to: [209756860](https://github.com/dotnet/machinelearning/pull/653#discussion_r209756860) [](ancestors = 209756860)

Reply:
I want to not expose MadeObjectPool

---
In reply to: [209768067](https://github.com/dotnet/machinelearning/pull/653#discussion_r209768067) [](ancestors = 209768067,209756860)

Reply:
Actually, I don't want any additional classes, just use the existing PredictionEngine. Which means that you just do `Parallel.ForEach(data, x=> lock(lockObject { engine.Predict(x); } ;

---
In reply to: [209770093](https://github.com/dotnet/machinelearning/pull/653#discussion_r209770093) [](ancestors = 209770093,209768067,209756860)